gocse/helpers/azure.go

```diff
@@ -0,0 +1,67 @@
+package helpers
+
+import (
+	"log"
+	"os"
+
+	"github.com/fatih/structs"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+// KMS Providers, the credentials used to access the KMS provider
+// See https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#kmsproviders
+type azureKMSCredentials struct {
+	TenantID                 string `structs:"tenantId"`
+	ClientID                 string `structs:"clientId"`
+	ClientSecret             string `structs:"clientSecret"`
+	IdentityPlatformEndpoint string `structs:"identityPlatformEndpoint,omitempty"` // optional, defaults to login.microsoftonline.com
+}
+
+// KMS data key options.
+// See https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#datakeyopts
+type azureKMSDataKeyOpts struct {
+	KeyVaultEndpoint string `bson:"keyVaultEndpoint"`
+	KeyName          string `bson:"keyName"`
+	KeyVersion       string `bson:"keyVersion,omitempty"` // optional, defautls to the key's primary version
+}
+
+type Azure struct {
+	credentials azureKMSCredentials
+	dataKeyOpts azureKMSDataKeyOpts
+}
+
+func (az *Azure) Credentials() map[string]interface{} {
+	return structs.Map(az.credentials)
+}
+
+func (az *Azure) DataKeyOpts() []byte {
+	data, err := bson.Marshal(az.dataKeyOpts)
+	if err != nil {
+		log.Fatalf("couldn't marshal azure data key opts %v", err)
+	}
+	return data
+}
+func GetAzure() *Azure {
+	// azure kms provider credentials
+	azureTenantID := os.Getenv("FLE_AZURE_TENANT_ID")
+	azureClientID := os.Getenv("FLE_AZURE_CLIENT_ID")
+	azureClientSecret := os.Getenv("FLE_AZURE_CLIENT_SECRET")
+
+	// azure data key opts
+	azureKeyVaultEndpoint := os.Getenv("FLE_AZURE_KEYVAULT_ENDPOINT")
+	azureKeyName := os.Getenv("FLE_AZURE_KEY_NAME")
+
+	return &Azure{
+		credentials: azureKMSCredentials{
+			azureTenantID,
+			azureClientID,
+			azureClientSecret,
+			"", // optional param we aren't using
```

Comment:
If optional, why do we need an empty string there?

Reply:
I did it for struct initialization, but it isn't included in the bson or map

Reply:
Oh ok, I just don't like having empty params when I can help it. I guess the the verbosity could be distracting if you had pursued one of the following alternatives:
- created an overridden constructor for only the required arguments and another for the one with the optional param
- used pointers to let you declare an arbitrary number of params https://willnorris.com/2014/05/go-rest-apis-and-pointers/
- marshal json/bson directly into the struct

Reply:
These are now removed